### ISSUE TITLE
Instance memory usage question

### ISSUE THREAD

**Issue by BenHayat** [DevExpress/Blazor#26]
Hello team;

Have you guys done any memory usage testing when an instance of a Blazor app gets instantiated?
For example, opening multiple chrome tabs, i.e. 20 tabs and each time you open a tab and call the same URL (which causes on the server to instantiate a new copy of the app), which basically replicates as users are using the app.
From different tests (running on my dev machine that is running the app and VS19), and watching the total memory usage. Depending on the size of the app, but for each user it's anywhere from 20 MB RAM to 40 MB RAM increase. The DX Demo was going up about 40 MB per tab.

Which means you can get limited very fast in a public facing app.
Have you guys done some rough benchmark testing on CPU and Memory usage as # of users increase?

Thanks!
..Ben

**Comment by SergeySeleznyov** [comment]
Hello Ben,

I am concerned to hear that our calculation result may mismatch the customer's real result.
Since a server-side Blazor application consumes some memory per active user, the total amount of consumed resources should be proportional to:

      ActiveUserNumber * ResourceAmountConsumedPerUser

But, since "ResourceAmountConsumedPerUser" heavily depends on the details of each application, the numbers may be unpredictable in practice.
 
From the developer's point of view, this means that it makes sense to not copy/store large data chunks in page fields.
We design components in the way they work with partial data. Since components are destroyed when a user leaves the page, we do not lock the data from GC.
The "services" that are registered in Startup.cs may help to control the data life time. For instance, we can use AddSingleton for every constant data source like "Countries".
 
As a result, we can do some calculation, for instance, for our demo, but I'm not sure if the calculation result will make sense outside our demos.
 
It is a common difficulty for stateful applications. We worried about this when we developed our stateful ASP.NET ASPxSpreadsheet/ASPxRichEdit controls. They may consume more than 40 MB for complex documents. Any user can open any number of documents, so the session size may vary greatly as well.
There are some practices how to reduce the RAM usage. For instance, in our ASP.NET Office controls we developed "Hybernation" when a session is not active for some time. Then it moves from RAM to some storage (a file system). There are strategies to release outdated sessions.
Thus having used such practices in ASP.NET, we think that there won't lead to any problem for most of developers.
For more complex cases it makes sense to develop distributed application. 

Microsoft declared that Blazor can work with Azure. They use their SignalR Hub for "Sticky session" implementation. So even disconnected users will be connected to the proper server (that holds this user's session). In the current state of the life cycle, Blazor changes too rapidly, so I really think that they may be changing something while I'm writing an answer to you. Moreover, the Microsoft Blazor Preview 7 update is right around the corner. 
To ensure that you have all the actual information, please ask the Blazor team about their current session releasing strategy. The guys there are awesome and answer all my question very fast.

I hope I managed to help.
Thank you, Ben, for your continued interest.

**Comment by BenHayat** [comment]
Hi Sergey;
I was just sharing my findings with the team as what might be up against with server-side blazor. In the old days of "Web Form" we used to see those large session data. But I think, Blazor is one step worst than that.
Every new user that starts Blazor App the first time, a NEW copy of that App gets created with it's own SignalR server side data and etc. Just like the old Terminal services, which makes sense. The numbers I was getting, of course applies to my case, my app, my settings and so on. But what alerted me (which I was expecting), was that each tab that opened a new user, the memory usage went up, which basically tells is Blazor Server side isn't meant for large public facing app, like if we were running a Web API to respond to an Angular or Vue client, that only call the server when they need data.

I just thought I share this, so other developers don't get under illusion or blame DX components. It is Blazor's model.

> To ensure that you have all the actual information, please ask the Blazor team about their current session releasing strategy.

I actually raised that question on Blazor Issues, which was closed as a subject that Microsoft is planning to address that later with docs. Dan Roth responded that it be covered in the docs. 

Basically, that tells me, I have to be careful how I publish my app until the Client-Side version comes out which should NOT have this problem. It was good test for me, not to make big size plans with lots of users in mind (which I was going that direction at first) and now I know, I have a tool (Blazor) can run out of breath quickly. :-)

Thank you as always responding to richly to cover and educate us.
..Ben

**Comment by BenHayat** [comment]
You may close it or leave it open for others to read.

**Comment by SergeySeleznyov** [comment]
Yes, holding user states on the server(s) is the price but let’s see what you pay for: 
Consider some abstract request:

ASP.NET:
- Sends Viewstate and all input values (actually the app state for this user) from the client to the server (can be several MB sometimes)
- Creates a new instance of the asp.net page
- Creates a hierarchy of all the controls on the page and initializes them
- Applies ViewState (and ControlState)
- There should be data binding somewhere here (depends on the app details) to properly accept the "Post data"
- Applies the "Post data" (all the input values even if they are not necessary for this request)
- Differences between the ViewState and PostData control states trigger custom "...Changed" event handlers -> custom code changes page/controls state
- There should be another data binding somewhere here based on the new page/controls state changed by custom event handlers (depends on the app details)
...
- Persists a new ViewState for future requests
- Responds to the client with a new render and ViewState
- Disposes of the page and control hierarchy
 
Actually, this lifecyce is simplified and generalized.
 
Blazor:
- Sends a message about a user action (mouse click, input value changed, etc.) to the c# code (to the server in the case of a server-side Blazor, or locally in the case of a client-side Blazor).
- Processes some custom code (since the state is kept up-to-date)
- Tracks what has changed
- Returns the virtual DOM difference to the client code (from the server to the client in the case of a server-side Blazor, or locally in the case of a client-side Blazor)
- Just to make this list look longer: applies the DOM changes to the HTML in the client code.
 
So, while ASP.NET recreates the whole App part the user communicates with (page + controls + catches up their states) for each server-side action, the Blazor is ready to process a user action at any moment because it keeps the client state up to date.
 
For example: there is a Grid with some row being edited and there is an opened ComboBox inside that edited row.
Scroll the ComboBox list to load the next portion of combobox items from the server.
 
 
Blazor does not recreate the whole page/controls/states; it keeps the state up-to-date: it keeps the Grid instance with proper data already loaded, with a row in the editing state, with the combobox inside, and the combobox is opened and ready to process the user's scroll action once it receives any.
 
### If you have more users a single server can handle, you can develop a distributed application.
 
For ASP.NET this will increase the number of these long lifecycles you can process in parallel.
For a server-side Blazor it will increase the number of states you can keep for users active at the same moment of time?
 
But:
1. Blazor's request is shorter, because:
  - there is no ViewState/ControlState
  - there are no unnecessary input values (ASP.NET sends all of them because only the server knows which of them matter).
  (in the case of a client-side Blazor, the request is local)
 
2. Blazor's request processing list of tasks looks shorter - see the lists above.*
 
> \* of course, to be precise we need some benchmark results to say which component is faster. Currently, we have no similar Blazor-based components to compare with, and Blazor is not as mature as ASP.NET. So, even if we had those benchmark results, not sure we could compare them. That is why I say "the list of tasks is shorter". We can image what task list should be completed to process requests on both of the platforms. I think currently the "task list" comparison is a more accurate way to estimate the performance difference.
 
3. Blazor's response is shorter:
  - even on a "callback" ASP.NET sends the whole render for the updated part of the page, while Blazor sends only the difference.  
  (the response is local in the case of a client-side Blazor)
 
So, holding user states on the server(s) is the price. Of course, there is no silver bullet here, but looking at the text above I can image the situations when it makes sense to pay.

**Comment by BenHayat** [comment]
@SergeySeleznyov ;
Thank you for well defined reply;
The information you have provided will help us a lot to determine the proper framework for our app.

What makes things more complex, is the Mobile platform, especially when user is not not on WiFi and needs speedy response from the app. The initial load time is very critical, so if the app becomes huge like with .Net runtime and Web Assembly runtime, it hurts the app.
If user gets turned off by slow loading, we basically lose that user, which is income to the app. Although most of the initial setup for a user or vendor in our app, will be done on Desktop browser which they have plenty of speed, but 95% of daily usage will be done by consumers via phone. And this is where we have to really find out the best solution for.

For large # of users, Blazor Server Side is not a good choice, if the # of users surges.
For Mobile size, with current MONO and WASM added to the app size is also not a good solution. We have to see how .Net 5 will be. These are realities that as solution provider we face...
But I'm glad we have DX team behind us. :-)

**Comment by d668** [comment]
>Microsoft declared that Blazor can work with Azure. 

Azure prices are ridiculous and I advice everyone to stay away from it and just buy a VM or a dedicated server.  They want 50$ a month for a website capable to have a SSL cert.
